File: wit-wpt-interop/ts/src/scenarios.test.ts
// EXPERIMENTAL: exploratory WIT/WPT interop study artifact (2026-09).
// 探索性试验产物，非稳定接口，勿用于生产或作为规范依据。

// S1..S11 场景矩阵（与 go/scenarios_test.go 一一对应，同一表格编号）。
// 运行：
//   node --disable-warning=ExperimentalWarning --experimental-strip-types --test src/scenarios.test.ts

import { test } from "node:test";
import assert from "node:assert/strict";
import { createPublicKey, randomBytes } from "node:crypto";
import type { KeyObject } from "node:crypto";

import { generateES256Key, keyHashOf } from "./jwk.ts";
import { Issuer, signArtifact } from "./issuer.ts";
import { Verifier, mintWPT, MemoryReplayStore, witKeyOf } from "./verifier.ts";
import { capabilityCoveredIn } from "./subset.ts";
import { signJWS, b64e } from "./jose.ts";
import type {
  Capability,
  DA,
  PA,
  Principal,
  Request,
  VerifyOpts,
  WITGrant,
  Decision,
} from "./types.ts";

const now0 = 1_700_000_000;
const targetURI = "https://rs.example.com/orders";

class Env {
  principal: KeyObject;
  issuerKey: KeyObject;
  iss: Issuer;
  ver: Verifier;

  constructor() {
    this.principal = generateES256Key();
    this.issuerKey = generateES256Key();
    this.iss = new Issuer("example.org", this.issuerKey, "wk-01");
    this.ver = new Verifier(
      { "wk-01": createPublicKey(this.issuerKey) },
      new MemoryReplayStore(),
    );
    this.ver.nowFn = () => now0;
  }
}

function mustKey(t: { name: string }): KeyObject {
  void t;
  return generateES256Key();
}

/** 构造 Capability，params 经 JSON 规范化（数字统一成 number）。 */
function capOf(scheme: string, id: string, params?: unknown): Capability {
  const c: Capability = { scheme, id };
  if (params !== undefined) {
    c.params = JSON.parse(JSON.stringify(params));
  }
  return c;
}

function principalOf(priv: KeyObject, realm: string, id: string): Principal {
  return {
    realm,
    id,
    key_hash: keyHashOf(createPublicKey(priv), "sha-256"),
    hash_alg: "sha-256",
  };
}

function makeDA(
  p: Principal,
  agentID: string,
  caps: Capability[],
  mode: DA["delegation_mode"],
  lifetime: number,
  now: number,
): DA {
  return {
    ver: 1,
    agent_id: agentID,
    principal: p,
    reason: { code: "access", desc: "grid access granted" },
    capabilities: caps,
    delegation_mode: mode,
    constraints: [],
    requested_lifetime: lifetime,
    ts: now,
    nonce: b64e(randomBytes(32)),
  };
}

function makePA(p: Principal, grants: Capability[]): PA {
  return { ver: 1, principal: p, grants, constraints: [] };
}

function daCompact(env: Env, d: DA): string {
  return signArtifact("da+jwt", d, env.principal);
}

function paCompact(env: Env, p: PA): string {
  return signArtifact("pa+jwt", p, env.principal);
}

/** 用发放服务密钥对修改后的 claims 重新签名（测试专用：WIT "改了字节但仍是有效生产产物"）。 */
function reSignWIT(env: Env, wit: string, mutate: (c: Record<string, unknown>) => void): string {
  const parts = wit.split(".");
  const header = JSON.parse(Buffer.from(parts[0], "base64url").toString());
  const claims = JSON.parse(Buffer.from(parts[1], "base64url").toString());
  mutate(claims);
  return signJWS(header, claims, env.issuerKey);
}

function happyGrant(env: Env): WITGrant {
  const p = principalOf(env.principal, "realm.example", "alice");
  const grants = [capOf("mysql", "query", { max_rows: 1000 })];
  const caps = [capOf("mysql", "query", { max_rows: 500 })];
  const da = makeDA(p, "agent-1", caps, "representative", 3600, now0);
  const pa = makePA(p, grants);
  return env.iss.issue(daCompact(env, da), paCompact(env, pa), "agent-1", now0);
}

function entitlementPolicy() {
  return (_ent: Capability[], c: Capability): void => {
    const { ok } = capabilityCoveredIn(c, _ent);
    if (!ok) {
      throw new Error("requested capability not in entitlement");
    }
  };
}

function happyInputs(env: Env) {
  const g = happyGrant(env);
  const wpt = mintWPT(g.wit_svid, "GET", targetURI, now0, witKeyOf(g));
  const rc = capOf("mysql", "query", { max_rows: 500 });
  const req: Request = {
    method: "GET",
    target_uri: targetURI,
    now: now0,
    requested_capability: rc,
  };
  const opts: VerifyOpts = { entitlement: g.entitlement, policy: entitlementPolicy() };
  return { wit: g.wit_svid, wpt, req, opts };
}

interface Run {
  wit: string;
  wpt: string;
  req: Request;
  opts: VerifyOpts;
  issueErr: string;
}

interface Case {
  id: string;
  name: string;
  wantPermit: boolean;
  wantReason: string;
  prepare: (env: Env) => Run;
}

const cases: Case[] = [
  {
    id: "S1",
    name: "happy path",
    wantPermit: true,
    wantReason: "",
    prepare: (env) => {
      const r = happyInputs(env);
      return { ...r, issueErr: "" };
    },
  },
  {
    id: "S2",
    name: "agent caps exceed P_grants (max_rows 5000 > 1000)",
    wantPermit: false,
    wantReason: "subset",
    prepare: (env) => {
      const p = principalOf(env.principal, "realm.example", "alice");
      const grants = [capOf("mysql", "query", { max_rows: 1000 })];
      const caps = [capOf("mysql", "query", { max_rows: 5000 })];
      const da = makeDA(p, "agent-1", caps, "representative", 3600, now0);
      const pa = makePA(p, grants);
      let issueErr = "";
      try {
        env.iss.issue(daCompact(env, da), paCompact(env, pa), "agent-1", now0);
      } catch (e) {
        issueErr = (e as Error).message;
      }
      return { wit: "", wpt: "", req: {} as Request, opts: {}, issueErr };
    },
  },
  {
    id: "S3",
    name: "tampered WIT + original WPT -> wth mismatch",
    wantPermit: false,
    wantReason: "wth mismatch",
    prepare: (env) => {
      const r = happyInputs(env);
      const bad = reSignWIT(env, r.wit, (c) => (c["tamper_flag"] = true));
      return { wit: bad, wpt: r.wpt, req: r.req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S4",
    name: "WPT signed with a different key",
    wantPermit: false,
    wantReason: "pop",
    prepare: (env) => {
      const r = happyInputs(env);
      const other = mustKey({ name: "S4" });
      const g = happyGrant(env);
      const badWpt = mintWPT(g.wit_svid, "GET", targetURI, now0, other);
      return { wit: g.wit_svid, wpt: badWpt, req: r.req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S5",
    name: "WPT aud != request target URI",
    wantPermit: false,
    wantReason: "aud mismatch",
    prepare: (env) => {
      const r = happyInputs(env);
      const req: Request = {
        method: "POST",
        target_uri: "https://rs.example.com/invoices",
        now: now0,
      };
      return { wit: r.wit, wpt: r.wpt, req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S6",
    name: "WPT expired (now > exp)",
    wantPermit: false,
    wantReason: "expired",
    prepare: (env) => {
      const r = happyInputs(env);
      env.ver.nowFn = () => now0 + 121; // WPT exp = now0+120
      const req: Request = { ...r.req, now: now0 + 121 };
      return { wit: r.wit, wpt: r.wpt, req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S7",
    name: "WPT replay",
    wantPermit: false,
    wantReason: "replay",
    prepare: (env) => {
      const r = happyInputs(env);
      const first: Decision = env.ver.verify(r.wit, r.wpt, r.req, r.opts);
      assert.equal(first.permit, true, `first verify should be PERMIT: ${first.reason}`);
      return { ...r, issueErr: "" };
    },
  },
  {
    id: "S8",
    name: "no WPT (bearer misuse)",
    wantPermit: false,
    wantReason: "pop required",
    prepare: (env) => {
      const r = happyInputs(env);
      env.ver.replay = new MemoryReplayStore();
      return { wit: r.wit, wpt: "", req: r.req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S9",
    name: "WIT carries aud (WIT-SVID MUST NOT)",
    wantPermit: false,
    wantReason: "aud-in-wit",
    prepare: (env) => {
      const r = happyInputs(env);
      const bad = reSignWIT(env, r.wit, (c) => (c["aud"] = targetURI));
      return { wit: bad, wpt: r.wpt, req: r.req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S10",
    name: "requested capability outside entitlement -> policy deny",
    wantPermit: false,
    wantReason: "policy:",
    prepare: (env) => {
      const r = happyInputs(env);
      const rc = capOf("mysql", "drop");
      const req: Request = { method: "POST", target_uri: targetURI, now: now0, requested_capability: rc };
      return { wit: r.wit, wpt: r.wpt, req, opts: r.opts, issueErr: "" };
    },
  },
  {
    id: "S11",
    name: "WPT alg != WIT cnf.jwk.alg",
    wantPermit: false,
    wantReason: "alg mismatch",
    prepare: (env) => {
      const r = happyInputs(env);
      const g = happyGrant(env);
      const valid = mintWPT(g.wit_svid, "GET", targetURI, now0, witKeyOf(g));
      const parts = valid.split(".");
      const payload = JSON.parse(Buffer.from(parts[1], "base64url").toString());
      const bad = signJWS({ typ: "wpt+jwt", alg: "RS256" }, payload, witKeyOf(g));
      const req: Request = { method: "GET", target_uri: targetURI, now: now0 };
      return { wit: g.wit_svid, wpt: bad, req, opts: r.opts, issueErr: "" };
    },
  },
];

for (const c of cases) {
  test(`${c.id} ${c.name}`, () => {
    const env = new Env();
    const run = c.prepare(env);
    if (run.issueErr !== "") {
      assert.equal(c.wantPermit, false, `expected issuance OK, got error ${run.issueErr}`);
      assert.ok(
        run.issueErr.includes(c.wantReason),
        `issue error "${run.issueErr}" missing "${c.wantReason}"`,
      );
      return;
    }
    const d = env.ver.verify(run.wit, run.wpt, run.req, run.opts);
    assert.equal(d.permit, c.wantPermit, `reason=${d.reason} checks=${JSON.stringify(d.checks)}`);
    if (!d.permit) {
      assert.ok(
        d.reason.includes(c.wantReason),
        `reason "${d.reason}" missing expected "${c.wantReason}"`,
      );
    }
  });
}

File: wit-wpt-interop/go/scenarios_test.go
// EXPERIMENTAL: exploratory WIT/WPT interop study artifact (2026-09).
// 探索性试验产物，非稳定接口，勿用于生产或作为规范依据。

package interop

import (
	"crypto/ecdsa"
	"crypto/rand"
	"encoding/json"
	"errors"
	"strings"
	"testing"
)

// env 组装一次完整场景所需的密钥与组件（每行独立、不共享状态）。
type env struct {
	principal *ecdsa.PrivateKey
	issuerKey *ecdsa.PrivateKey
	iss       *Issuer
	ver       *Verifier
}

func newEnv(t *testing.T) *env {
	t.Helper()
	pk := mustKey(t)
	ik := mustKey(t)
	iss := NewIssuer("example.org", ik, "wk-01")
	ver := &Verifier{
		TrustBundle: map[string]*ecdsa.PublicKey{"wk-01": &ik.PublicKey},
		Replay:      NewMemoryReplayStore(),
		Now:         func() int64 { return now0 },
	}
	return &env{principal: pk, issuerKey: ik, iss: iss, ver: ver}
}

func mustKey(t *testing.T) *ecdsa.PrivateKey {
	t.Helper()
	k, err := generateKey()
	if err != nil {
		t.Fatal(err)
	}
	return k
}

// capOf 构造 Capability，params 经 JSON 规范化（数字统一成 float64）。
func capOf(scheme, id string, params any) Capability {
	c := Capability{Scheme: scheme, ID: id}
	if params == nil {
		return c
	}
	b, _ := json.Marshal(params)
	_ = json.Unmarshal(b, &c.Params)
	return c
}

// principalOf 由密钥派生 principal 的 key_hash。
func principalOf(priv *ecdsa.PrivateKey, realm, id string) Principal {
	kh, err := keyHash(&priv.PublicKey, "sha-256")
	if err != nil {
		panic(err)
	}
	return Principal{Realm: realm, ID: id, KeyHash: kh, HashAlg: "sha-256"}
}

func makeDA(p Principal, agentID string, caps []Capability, mode string, lifetime int, now int64) DA {
	nonce := make([]byte, 32)
	if _, err := rand.Read(nonce); err != nil {
		panic(err)
	}
	return DA{
		Ver:               1,
		AgentID:           agentID,
		Principal:         p,
		Reason:            Reason{Code: "access", Desc: "grid access granted"},
		Capabilities:      caps,
		DelegationMode:    mode,
		RequestedLifetime: lifetime,
		TS:                now,
		Nonce:             b64e(nonce),
	}
}

func makePA(p Principal, grants []Capability) PA {
	return PA{Ver: 1, Principal: p, Grants: grants}
}

func (e *env) daCompact(t *testing.T, d DA) string {
	t.Helper()
	s, err := signArtifact("da+jwt", d, e.principal)
	if err != nil {
		t.Fatal(err)
	}
	return s
}

func (e *env) paCompact(t *testing.T, p PA) string {
	t.Helper()
	s, err := signArtifact("pa+jwt", p, e.principal)
	if err != nil {
		t.Fatal(err)
	}
	return s
}

// reSignWIT 用发放服务密钥对修改后的 claims 重新签名（测试专用：模拟"改了字节
// 但仍是有效生产产物"，从而把 wth mismatch 暴露在验签之后）。
func (e *env) reSignWIT(t *testing.T, wit string, mutate func(map[string]any)) string {
	t.Helper()
	h, p, _, err := parseCompact(wit)
	if err != nil {
		t.Fatal(err)
	}
	var claims map[string]any
	if err := json.Unmarshal(p, &claims); err != nil {
		t.Fatal(err)
	}
	mutate(claims)
	tok, err := signJWS(h, mustJSON(claims), e.issuerKey)
	if err != nil {
		t.Fatal(err)
	}
	return tok
}

const (
	now0      = int64(1_700_000_000)
	targetURI = "https://rs.example.com/orders"
)

// happyGrant 构造 S1 的合法代表授权：DA max_rows=500，PA grants max_rows=1000。
func (e *env) happyGrant(t *testing.T) *WITGrant {
	t.Helper()
	p := principalOf(e.principal, "realm.example", "alice")
	grants := []Capability{capOf("mysql", "query", map[string]any{"max_rows": 1000})}
	caps := []Capability{capOf("mysql", "query", map[string]any{"max_rows": 500})}
	da := makeDA(p, "agent-1", caps, "representative", 3600, now0)
	pa := makePA(p, grants)
	g, err := e.iss.Issue(e.daCompact(t, da), e.paCompact(t, pa), "agent-1", now0)
	if err != nil {
		t.Fatal(err)
	}
	return g
}

// happyInputs 返回 S1 的完整验证输入（WIT+WPT+Request+opts）。
func (e *env) happyInputs(t *testing.T) (string, string, Request, VerifyOpts) {
	t.Helper()
	g := e.happyGrant(t)
	witKey, err := jwkToPriv(g.WITPrivateKey)
	if err != nil {
		t.Fatal(err)
	}
	wpt, err := MintWPT(g.WITSVID, "GET", targetURI, now0, witKey)
	if err != nil {
		t.Fatal(err)
	}
	rc := capOf("mysql", "query", map[string]any{"max_rows": 500})
	req := Request{Method: "GET", TargetURI: targetURI, Now: now0, RequestedCapability: &rc}
	opts := VerifyOpts{
		Entitlement: g.Entitlement,
		Policy: func(ent []Capability, c Capability) error {
			ok, _ := capabilityCoveredIn(c, ent)
			if !ok {
				return errors.New("requested capability not in entitlement")
			}
			return nil
		},
	}
	return g.WITSVID, wpt, req, opts
}

func entitlementPolicy() func([]Capability, Capability) error {
	return func(ent []Capability, c Capability) error {
		ok, _ := capabilityCoveredIn(c, ent)
		if !ok {
			return errors.New("requested capability not in entitlement")
		}
		return nil
	}
}

type run struct {
	wit, wpt string
	req      Request
	opts     VerifyOpts
	// issueErr 非空 = 场景应在签发阶段被拒（如 S2）。
	issueErr string
}

func TestScenarios(t *testing.T) {
	cases := []struct {
		id, name   string
		wantPermit bool
		wantReason string
		prepare    func(t *testing.T, e *env) *run
	}{
		{"S1", "happy path", true, "", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			return &run{wit: wit, wpt: wpt, req: req, opts: opts}
		}},
		{"S2", "agent caps exceed P_grants (max_rows 5000 > 1000)", false, "subset", func(t *testing.T, e *env) *run {
			p := principalOf(e.principal, "realm.example", "alice")
			grants := []Capability{capOf("mysql", "query", map[string]any{"max_rows": 1000})}
			caps := []Capability{capOf("mysql", "query", map[string]any{"max_rows": 5000})}
			da := makeDA(p, "agent-1", caps, "representative", 3600, now0)
			pa := makePA(p, grants)
			_, err := e.iss.Issue(e.daCompact(t, da), e.paCompact(t, pa), "agent-1", now0)
			return &run{issueErr: errString(err)}
		}},
		{"S3", "tampered WIT + original WPT -> wth mismatch", false, "wth mismatch", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			bad := e.reSignWIT(t, wit, func(c map[string]any) { c["tamper_flag"] = true })
			return &run{wit: bad, wpt: wpt, req: req, opts: opts}
		}},
		{"S4", "WPT signed with a different key", false, "pop", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			other, _ := generateKey()
			bad, err := MintWPT(wit, "GET", targetURI, now0, other)
			if err != nil {
				t.Fatal(err)
			}
			_ = wpt
			return &run{wit: wit, wpt: bad, req: req, opts: opts}
		}},
		{"S5", "WPT aud != request target URI", false, "aud mismatch", func(t *testing.T, e *env) *run {
			wit, wpt, _, opts := e.happyInputs(t)
			req := Request{Method: "POST", TargetURI: "https://rs.example.com/invoices", Now: now0}
			return &run{wit: wit, wpt: wpt, req: req, opts: opts}
		}},
		{"S6", "WPT expired (now > exp)", false, "expired", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			e.ver.Now = func() int64 { return now0 + 121 } // WPT exp = now0+120
			req.Now = now0 + 121
			return &run{wit: wit, wpt: wpt, req: req, opts: opts}
		}},
		{"S7", "WPT replay", false, "replay", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			if d := e.ver.Verify(wit, wpt, req, opts); !d.Permit {
				t.Fatalf("first verify should be PERMIT: %v", d.Reason)
			}
			return &run{wit: wit, wpt: wpt, req: req, opts: opts}
		}},
		{"S8", "no WPT (bearer misuse)", false, "pop required", func(t *testing.T, e *env) *run {
			wit, _, req, opts := e.happyInputs(t)
			e.ver.Replay = NewMemoryReplayStore()
			return &run{wit: wit, wpt: "", req: req, opts: opts}
		}},
		{"S9", "WIT carries aud (WIT-SVID MUST NOT)", false, "aud-in-wit", func(t *testing.T, e *env) *run {
			wit, wpt, req, opts := e.happyInputs(t)
			bad := e.reSignWIT(t, wit, func(c map[string]any) { c["aud"] = targetURI })
			return &run{wit: bad, wpt: wpt, req: req, opts: opts}
		}},
		{"S10", "requested capability outside entitlement -> policy deny", false, "policy:", func(t *testing.T, e *env) *run {
			wit, wpt, _, opts := e.happyInputs(t)
			rc := capOf("mysql", "drop", nil)
			req := Request{Method: "POST", TargetURI: targetURI, Now: now0, RequestedCapability: &rc}
			return &run{wit: wit, wpt: wpt, req: req, opts: opts}
		}},
		{"S11", "WPT alg != WIT cnf.jwk.alg", false, "alg mismatch", func(t *testing.T, e *env) *run {
			wit, _, _, opts := e.happyInputs(t)
			g := e.happyGrant(t)
			witKey, err := jwkToPriv(g.WITPrivateKey)
			if err != nil {
				t.Fatal(err)
			}
			valid, _ := MintWPT(wit, "GET", targetURI, now0, witKey)
			_, p, _, err := parseCompact(valid)
			if err != nil {
				t.Fatal(err)
			}
			hdr := map[string]any{"typ": "wpt+jwt", "alg": "RS256"}
			bad, err := signJWS(mustJSON(hdr), p, witKey)
			if err != nil {
				t.Fatal(err)
			}
			req := Request{Method: "GET", TargetURI: targetURI, Now: now0}
			return &run{wit: wit, wpt: bad, req: req, opts: opts}
		}},
	}

	for _, c := range cases {
		t.Run(c.id+" "+c.name, func(t *testing.T) {
			e := newEnv(t)
			r := c.prepare(t, e)
			if r.issueErr != "" {
				if c.wantPermit {
					t.Fatalf("expected issuance OK, got error %q", r.issueErr)
				}
				expectContains(t, r.issueErr, c.wantReason)
				return
			}
			if c.wantReason == "" && c.wantPermit == false {
				t.Fatalf("test case must specify wantReason or wantPermit=true")
			}
			d := e.ver.Verify(r.wit, r.wpt, r.req, r.opts)
			if d.Permit != c.wantPermit {
				t.Fatalf("Permit=%v want %v; reason=%q checks=%+v",
					d.Permit, c.wantPermit, d.Reason, d.Checks)
			}
			if !d.Permit {
				expectContains(t, d.Reason, c.wantReason)
			}
		})
	}
}

func expectContains(t *testing.T, s, sub string) {
	t.Helper()
	if sub != "" && !strings.Contains(s, sub) {
		t.Fatalf("reason %q missing expected substring %q", s, sub)
	}
}

func errString(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}
